compiler/src/iree/compiler/Dialect/Flow/Conversion/MeshToFlow/MeshToFlow.cpp

```diff
@@ -0,0 +1,789 @@
+             meshAndAxesSetSorted.rbegin(), meshAndAxesSetSorted.rend())) {
+      buildGlobalChannelCreation(mesh, meshAxes, analysis.hasMoreThanOneMesh,
+                                 getOperation(), builder);
+    }
+
+    markAnalysesPreserved<MeshToFlowModuleOpAnalysis>();
+  }
+};
+
+struct ConvertMeshToFlowCollectivesPass
+    : public Flow::ConvertMeshToFlowCollectivesBase<
+          ConvertMeshToFlowCollectivesPass> {
+  void getDependentDialects(DialectRegistry &registry) const override {
+    registerMeshToFlowDependencies(registry);
+  }
+
+  template <typename AnalysisT>
+  std::optional<std::reference_wrapper<AnalysisT>>
+  getCachedAncestorAnalysis(Operation *op) {
+    if (op == getOperation())
+      return getCachedAnalysis<AnalysisT>();
+    else
+      return getCachedParentAnalysis<AnalysisT>(op);
+  }
+
+  void runOnOperation() override {
+    RewritePatternSet patterns(&getContext());
+    SymbolTableCollection symbolTableCollection;
+    Flow::populateMeshToFlowCollectivesPatterns(
+        patterns, symbolTableCollection,
+        getCachedAncestorAnalysis<MeshToFlowModuleOpAnalysis>(getOperation())
+            .value()
+            .get()
+            .hasMoreThanOneMesh);
+    if (failed(applyPatternsAndFoldGreedily(getOperation(),
+                                            std::move(patterns)))) {
+      return signalPassFailure();
+    }
+  }
+};
+
+struct RemoveMeshClusterOpsPass
+    : public RemoveMeshClusterOpsBase<RemoveMeshClusterOpsPass> {
+  void runOnOperation() override {
+    RewritePatternSet patterns(&getContext());
+    getOperation()->walk([](mesh::ClusterOp op) {
+      op->erase();
+      return WalkResult::skip();
+    });
+  }
+};
+
+} // namespace
+
+void populateMeshToFlowCollectivesPatterns(
+    RewritePatternSet &patterns, SymbolTableCollection &symbolTableCollection,
+    bool useNamedDefaultChannels) {
+  patterns.add<MeshAllGatherToFlow, MeshAllReduceToFlow, MeshAllToAllToFlow,
+               MeshReduceScatterToFlow, MeshProcessLinearIndexToFlow>(
+      symbolTableCollection, useNamedDefaultChannels, patterns.getContext());
+  mesh::populateFoldingPatterns(patterns, symbolTableCollection);
+  mesh::processMultiIndexOpLoweringPopulatePatterns(patterns,
+                                                    symbolTableCollection);
+}
+
+std::unique_ptr<Pass> createMeshToFlowChannelConstructionPass() {
+  return std::make_unique<MeshToFlowChannelConstructionPass>();
+}
+
+std::unique_ptr<Pass> createConvertMeshToFlowCollectivesPass() {
+  return std::make_unique<ConvertMeshToFlowCollectivesPass>();
+}
+
+std::unique_ptr<Pass> createRemoveMeshClusterOpsPass() {
+  return std::make_unique<RemoveMeshClusterOpsPass>();
+}
+
+void buildMeshToFlowConversionPassPipeline(OpPassManager &passManager) {
+  passManager.addPass(createMeshToFlowChannelConstructionPass());
+  passManager.addPass(createConvertMeshToFlowCollectivesPass());
+  passManager.addPass(createRemoveMeshClusterOpsPass());
```

Comment:
There's a cost to that as it adds 3 full-module verifies at the least. e.g. walking the entire module (tens of thousands of ops) to remove a handful of ops that you could have a pattern remove during applyPatternsAndFoldGreedily (or that you already know of because of the analysis) is a lot of overhead. Getting it down to one pass makes it easier to verify the before/after state. In general splitting the passes is useful if they are optional or reorderable or independently useful to test but in this case they aren't: you MUST run them in this sequence and run all of them in order to get valid IR. One pass that does the global creation and applies the patterns (including one which removes the ClusterOps) should be able to take care of all of this pretty quickly.

Reply:
Thank you for the suggestion. I moved everything into one pass.